go/libraries/doltcore/sqle/autoincr/auto_increment_tracker.go

```diff
@@ -0,0 +1,115 @@
+package autoincr
+
+import (
+	"fmt"
+	"sync"
+)
+
+type AutoIncrementTracker interface {
+	// Next returns the next auto increment value to be used by a table. If a table is not initialized in the counter
+	// it will used the value stored in disk.
+	Next(tableName string, insertVal interface{}, diskVal interface{}) (interface{}, error)
+	// Reset resets the auto increment tracker value for a table. Typically used in truncate statements.
+	Reset(tableName string, val interface{})
+}
+
+// AutoIncrementTracker is a global map that tracks which auto increment keys have been given for each table. At runtime
+// it hands out the current key.
+func NewAutoIncrementTracker() AutoIncrementTracker {
+	return &autoIncrementTracker{
+		valuePerTable: make(map[string]interface{}),
+	}
+}
+
+type autoIncrementTracker struct {
+	valuePerTable map[string]interface{}
+	mu            sync.Mutex
+}
+
+var _ AutoIncrementTracker = (*autoIncrementTracker)(nil)
+
+func (a *autoIncrementTracker) Next(tableName string, insertVal interface{}, diskVal interface{}) (interface{}, error) {
```

Comment:
Do me a favor and write a unit test that spins up 100 goroutines that all call Next with nil 10 times, and asserts that none of them get the same value back